File: backend/api/routes/catalog.py
"""GET /catalog — the published workflow catalog with version supersession.

`superseded_by` is derived by the filename convention (a version is a file): strip a trailing `_v{n}` to get the family stem (no suffix == v1);
the highest version in a family is current, every other member is
superseded_by the highest."""

from __future__ import annotations

import re

from fastapi import APIRouter, Depends

from api.deps import get_db
from engine import db as dbm

router = APIRouter(tags=["catalog"])

_VERSION_RE = re.compile(r"^(.+)_v(\d+)$")


def _family(workflow_id: str) -> tuple[str, int]:
    m = _VERSION_RE.match(workflow_id)
    return (m.group(1), int(m.group(2))) if m else (workflow_id, 1)


def superseded_map(workflow_ids: list[str]) -> dict[str, str | None]:
    families: dict[str, list[tuple[int, str]]] = {}
    for wid in workflow_ids:
        stem, version = _family(wid)
        families.setdefault(stem, []).append((version, wid))
    out: dict[str, str | None] = {}
    for members in families.values():
        current = max(members)[1]
        for _, wid in members:
            out[wid] = None if wid == current else current
    return out


@router.get("/catalog")
def get_catalog(conn=Depends(get_db)) -> dict:
    snapshot = dbm.catalog_snapshot(conn)
    superseded = superseded_map([w["workflow_id"] for w in snapshot])
    return {
        "workflows": [
            {
                "workflow_id": w["workflow_id"],
                "display_name": w["display_name"],
                "task_queue": w["task_queue"],
                "superseded_by": superseded[w["workflow_id"]],
                "kinds": [
                    {"kind": k["kind"], "display_name": k["display_name"], "leaf": bool(k["leaf"])}
                    for k in w["kinds"]
                ],
                "nodes": [
                    {
                        "node_id": n["node_id"],
                        "display_name": n["display_name"],
                        "executor": n["executor"],
                        "output_kind": n["output_kind"],
                        "code_hash": n["code_hash"],
                    }
                    for n in w["nodes"]
                ],
            }
            for w in snapshot
        ]
    }

File: backend/src/api/routes/Catalog.ts
import type { FastifyInstance } from 'fastify';
import { catalogSnapshot } from '../../infrastructure/db/Db.js';
import type { ApiDeps } from '../Deps.js';
import { withConn } from '../Deps.js';
import type { CatalogOut } from '../Schemas.js';

const VERSION_RE = /^(.+)_v(\d+)$/;

// Strip a trailing _v{n}; no suffix == version 1.
function family(workflowId: string): { stem: string; version: number } {
  const m = VERSION_RE.exec(workflowId);
  if (m === null) {
    return { stem: workflowId, version: 1 };
  }
  return { stem: m[1], version: Number(m[2]) };
}

// The highest (version, workflow_id) member of each family gets superseded_by: null; every other
// member points at it. Derived purely from the workflow_id naming convention (the id is the
// workflow's folder name under src/workflows/).
function supersededMap(workflowIds: readonly string[]): Record<string, string | null> {
  const families = new Map<string, { version: number; wid: string }[]>();
  for (const wid of workflowIds) {
    const { stem, version } = family(wid);
    const members = families.get(stem) ?? [];
    members.push({ version, wid });
    families.set(stem, members);
  }
  const out: Record<string, string | null> = {};
  for (const members of families.values()) {
    let current = members[0];
    for (const m of members) {
      if (m.version > current.version || (m.version === current.version && m.wid > current.wid)) {
        current = m;
      }
    }
    for (const m of members) {
      out[m.wid] = m.wid === current.wid ? null : current.wid;
    }
  }
  return out;
}

export function registerCatalogRoutes(app: FastifyInstance, deps: ApiDeps): void {
  app.get('/catalog', async (): Promise<CatalogOut> => {
    const snapshot = withConn(deps, (conn) => catalogSnapshot(conn));
    const superseded = supersededMap(snapshot.map((wf) => wf.workflow_id));
    return {
      workflows: snapshot.map((wf) => ({
        workflow_id: wf.workflow_id,
        display_name: wf.display_name,
        created_at: wf.created_at,
        updated_at: wf.updated_at,
        superseded_by: superseded[wf.workflow_id] ?? null,
        nodeparamslots: wf.nodeparamslots.map((k) => ({
          nodeparamslot: k.nodeparamslot,
          display_name: k.display_name,
          source: k.source,
          leaf: k.leaf !== 0,
        })),
        nodes: wf.nodes.map((n) => ({
          node_id: n.node_id,
          display_name: n.display_name,
          executor: n.executor,
          output_nodeparamslot: n.output_nodeparamslot,
          input_nodeparamslots: n.input_nodeparamslots,
        })),
      })),
    };
  });
}
